### DISCUSSION TITLE
Correct way to access modules that aren't part of the dependencies in PnP strict mode

### DISCUSSION THREAD

**Discussion by Blackclaws** [yarnpkg/berry#4940] [Q&A]
We have a node package that is a helper script that parses values in package.json and extracts files from the modules in question to assist in the build process. We use require.resolve("module") to get access to the on disk path of the module.

With PnP in strict mode the access to other modules is denied because the helper script doesn't directly declare it as a dependency.
Is there any way without switching everything to loose mode to get rid of the requirement to list modules in dependencies for scripts like this?

The require call in this case isn't unsound or ambiguous because the required modules in question are declared in the main projects dependencies. 

One possible solution I've thought of is writing a yarn plugin. I'm just wondering whether that's the _only_ officially supported solution.

**Comment by Blackclaws** [comment]
So what I've arrived at so far is using the pnpapi directly to resolve packages in case it is available. That allows me to distinguish whether I am running in pure npm or within yarn PnP. I can then use PnP's resolveRequest to resolve to the actual fileSystem location. That seems to work so far. If anyone else has a better idea that isn't so roundabout please let me know.

The solution here is to check if ```require('pnpapi')``` works. If it does you need to require fs as the yarnpkg version of fs see: https://yarnpkg.com/advanced/pnpapi#accessing-the-files

You can then get the file locations equivalent by running: 
```
packagePath = path.dirname(pnp.resolveRequest(packageName + "/package.json", process.env.npm_package_json));
```

**Comment by merceyz** [comment]
Use the `paths` option of `require.resolve`
```js
require.resolve('packageName', {paths: ['/path/to/resolve/from']})
```
https://nodejs.org/dist/latest-v16.x/docs/api/modules.html#requireresolverequest-options
